docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,281 @@
+
+Example:
+
+On Kubernetes, all tags are populated automatically, while on Universal, users need to populate tags themselves:
+
+```yaml
+type: Dataplane
+mesh: default
+name: demo
+networking:
+  address: 127.0.0.1
+  inbound:
+    - port: 9999
+      servicePort: 10000
+      serviceAddress: 127.0.0.1
+      tags:
+        kuma.io/service: demo
+        kuma.io/protocol: tcp
+        k8s.io/az: us-east-1
+```
+
+
+Later, based on these tags, the control-plane creates `ClusterLoadBalancing` and puts all tags, for each endpoint, into cluster metadata::
+
+```json
+"metadata":{
+  "filter_metadata": {
+    "envoy.lb": {
+      "az/test": "b",
+      "kuma.io/protocol": "tcp",
+      "k8s.io/az": "us-east-1"
+    }
+  }
+}
+```
+
+
+We can retrieve them from `ClusterLoadBalancing` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.
+
+Pseudo algorithm:
+
+1. Get all inbound tags.
+2. Take the matched rules.
+3. Check if inbound has the tag defined in the rule. 
+4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule( first inZone, then  cross Zone). If no, get all the endpoints in the zone. 
+5. Create `LocalityLbEndpoints` with selected endpoints. 
+6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceStore`. 
+
+Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.
+
+##### Egress
+Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress sidecar because there might be services sending requests to all zones. There are 2 options:
+- LoadBalacing based on metadata
+- Separate clusters
+
+Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.
+
+Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`
+This cluster has a different configuration of priorities and only subset of zones. Dataplanes interested in this Load Balancing are going to have the same configuration of the cluster which sets SNI: `service-name{mesh=mesh-name,hash=hash}` and based in this SNI egress can make a routing decision to route to the specific cluster.
```

Comment:
Could we just say that Egress will be the equivalent of top level targetRef be mesh and won't make a difference between client?

It's likely an acceptable shortcoming. I think down the line @jakubdyszkiewicz mentioned maybe having one cluster per zone (and route) which would make these things a bit easier maybe.

Reply:
Yes, cluster/route per zone would be simpler in this case because we could deliver just these clusters.